[quant] fix dropout in FX graph mode quantization

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71043

Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33490176](https://our.internmc.facebook.com/intern/diff/D33490176)

terrychenism committed:
[quant] fix dropout in FX graph mode quantization

Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
I think we probably don't need qat module, the support is similar to quantized batchnorm. we can add a test similar to batchnorm: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_module.py#L611 as well in the same TestCase

jerryzh168 commented:
Also would be good to include the test in the issue as well, we can add a test function in https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L5536

terrychenism committed:
Update on "[quant] fix dropout in FX graph mode quantization"


Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[quant] fix dropout in FX graph mode quantization"


Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

terrychenism committed:
Update on "[quant] fix dropout in FX graph mode quantization"


Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33490176](https://our.internmc.facebook.com/intern/diff/D33490176)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

terrychenism committed:
Update on "[quant] fix dropout in FX graph mode quantization"


Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33490176](https://our.internmc.facebook.com/intern/diff/D33490176)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

terrychenism committed:
Update on "[quant] fix dropout in FX graph mode quantization"


Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33490176](https://our.internmc.facebook.com/intern/diff/D33490176)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

terrychenism committed:
Update on "[quant] fix dropout in FX graph mode quantization"


Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33490176](https://our.internmc.facebook.com/intern/diff/D33490176)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

vadimkantorov commented on torch/nn/quantized/modules/dropout.py:
```diff
@@ -0,0 +1,26 @@
+    """
+    def __init__(self, p=0.5, inplace=False):
+        super(Dropout, self).__init__(p, inplace)
+        self.inplace = inplace
```
are these lines actually needed? the base class does set these fields, right? also, `super().__init__(p, inplace)` is a bit more modern / less verbose...

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -5617,6 +5617,15 @@ def forward(self, x):
                 out = model_quantized(input.to(device_after))
                 self.assertEqual(out.device.type, device_after)
 
+    def _test_model_dropout(self):
```
nit: this looks like a normal test instead of a helper function, probably don't need to start with `_`

jerryzh168 replied:
I think we can add another similar test in eager mode test as well: https://github.com/pytorch/pytorch/blob/master/test/quantization/eager/test_quantize_eager_qat.py#L246

jerryzh168 commented:
Looks good to me, but I'll let @vkuzo to take a final look

vkuzo commented on torch/nn/quantized/modules/dropout.py:
```diff
@@ -0,0 +1,26 @@
+import torch
+import torch.nn.quantized.functional
+
+class Dropout(torch.nn.Dropout):
```
Thanks for fixing this.  It would be valuable to add some context to why the logic of this class is not actually doing dropout, i.e. there is currently no need or implementation for actual quantized dropout, and this is a placeholder to enable models where fp32 tensors had dropout to work with quantized tensors.

vkuzo replied:
I'm also still not sure why not we just use `nn.Identity` instead of creating this module, but I don't feel super strongly

terrychenism replied:
I don't have any preference.  
@jerryzh168 any thoughts?

terrychenism committed:
Update on "[quant] fix dropout in FX graph mode quantization"


Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33490176](https://our.internmc.facebook.com/intern/diff/D33490176)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant] fix dropout in FX graph mode quantization"


Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33490176](https://our.internmc.facebook.com/intern/diff/D33490176)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant] fix dropout in FX graph mode quantization"


Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33490176](https://our.internmc.facebook.com/intern/diff/D33490176)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

terrychenism committed:
Update on "[quant] fix dropout in FX graph mode quantization"


Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33490176](https://our.internmc.facebook.com/intern/diff/D33490176)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

vkuzo commented on torch/nn/quantized/modules/dropout.py:
```diff
@@ -0,0 +1,22 @@
+import torch.nn.quantized.functional
+
+class Dropout(torch.nn.Dropout):
+    r"""This is the quantized equivalent of :class:`~torch.nn.Dropout`.
```
would be good to explicitly say "this class does not actually do dropout"

vkuzo reviewed (approved):
thanks for working on this!

datumbox commented:
I think this PR broke TorchVision's pre-trained quantized models. See here for all details: https://github.com/pytorch/vision/issues/5198
